streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreThread.java

```diff
@@ -0,0 +1,121 @@
+
+import org.apache.kafka.common.errors.TimeoutException;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.errors.TaskCorruptedException;
+import org.slf4j.Logger;
+
+import java.util.Collection;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+
+/**
+ * This is the thread responsible for restoring state stores for both active and standby tasks
+ */
+public class StateRestoreThread extends Thread {
+
+    private final Time time;
+    private final Logger log;
+    private final StreamsConfig config;
+    private final ChangelogReader changelogReader;
+    private final AtomicBoolean isRunning = new AtomicBoolean(true);
+    private final CountDownLatch shutdownLatch = new CountDownLatch(1);
+    private final ConcurrentLinkedDeque<TaskCorruptedException> corruptedExceptions;
+
+    public boolean isRunning() {
+        return isRunning.get();
+    }
+
+    public StateRestoreThread(final Time time,
+                              final StreamsConfig config,
+                              final String threadClientId,
+                              final ChangelogReader changelogReader) {
+        super(threadClientId);
+        this.time = time;
+        this.config = config;
+        this.changelogReader = changelogReader;
+        this.corruptedExceptions = new ConcurrentLinkedDeque<>();
+
+        final String logPrefix = String.format("state-restore-thread [%s] ", threadClientId);
+        final LogContext logContext = new LogContext(logPrefix);
+        this.log = logContext.logger(getClass());
+    }
+
+    @Override
+    public void run() {
+        try {
+            while (isRunning()) {
+                final long startMs = time.milliseconds();
+
+                try {
+                    // try to restore some changelogs, if there's nothing to restore it would wait inside this call
+                    final int numRestored = changelogReader.restore();
+                    // TODO: we should record the restoration related metrics
+                    log.info("Restored {} records in {} ms", numRestored, time.milliseconds() - startMs);
+                } catch (final TaskCorruptedException e) {
+                    log.warn("Detected the states of tasks " + e.corruptedTaskWithChangelogs() + " are corrupted. " +
+                            "Will close the task as dirty and re-create and bootstrap from scratch.", e);
+
+                    // remove corrupted partitions form the changelog reader and continue; we can still proceed
+                    // and restore other partitions until the main thread come to handle this exception
+                    changelogReader.unregister(e.corruptedTaskWithChangelogs().values().stream()
```

Comment:
I'm not sure this is strictly incorrect since unregistering is idempotent, but it certainly seems unnecessary. Why do we need to unregister the changelogs here? We do so in `ProcessorStateManager`'s `close` which should be called in `handleCorruption` from the main thread. 
I'm also generally against unregistering something so far away from where it gets registered 🙂

Reply:
I think that's a fair point --- all I care here is letting the changelogReader continue without needing to fetch on this partition any more, since the handling of the exception is on the main thread and hence async to this thread, we do not want to keep getting the same exception again and again.

Let me think how I can refactor this piece.